stonesoup/dataassociator/clearmot.py

```diff
@@ -0,0 +1,291 @@
+        # assication threshold - if true, we keep it and add it to current set,
+        # if not we do not maintain the match
+        for (track_id, truth_id) in matches_previous:
+            # get
+            truth_states = truth_states_by_id[truth_id]
+            truth_state_current = get_state_at_time(truth_states, current_time)
+
+            if not truth_state_current:
+                continue
+
+            track_states = track_states_by_id[track_id]
+            track_state_current = get_state_at_time(track_states, current_time)
+
+            # if hypothesis is not available anymore
+            if not track_state_current:
+                continue
+
+            distance = self.measure(track_state_current, truth_state_current)
+
+            # if distance is still lower than the threshold, keep the match
+            if distance < self.association_threshold:
+                matches_current.add((track_id, truth_id))
+
+                truth_ids_at_current_time.remove(truth_id)
+                track_ids_at_current_time.remove(track_id)
+        return matches_current
+
+    def match_unassigned_tracks(self, truth_states_by_id, track_states_by_id, current_time, truth_ids_at_current_time,
+                                track_ids_at_current_time, matches_current: Set):
+        num_truth_unassigned = len(truth_ids_at_current_time)
+        num_tracks_unassigned = len(track_ids_at_current_time)
+        cost_matrix = np.zeros((num_truth_unassigned, num_tracks_unassigned), dtype=float)
+
+        for i in range(num_truth_unassigned):
+            for j in range(num_tracks_unassigned):
+                truth_id, track_id = truth_ids_at_current_time[i], track_ids_at_current_time[j]
+
+                truth_states = truth_states_by_id[truth_id]
+                track_states = track_states_by_id[track_id]
+                truth_state_current = get_state_at_time(truth_states, current_time)
+                track_state_current = get_state_at_time(track_states, current_time)
+                distance = self.measure(track_state_current, truth_state_current)
+                cost_matrix[i, j] = distance
+
+        # Munkers / Hungarian Method for the assignment problem
+        row_ind, col_in = scipy.optimize.linear_sum_assignment(cost_matrix)
+
+        for i, j in zip(row_ind, col_in):
+            if cost_matrix[i, j] < self.association_threshold:
+                matches_current.add((track_ids_at_current_time[j], truth_ids_at_current_time[i]))
+
+    def determine_unique_timestamps(self, tracks_set: Set[Track], truth_set: Set[GroundTruthPath])\
+            -> list[datetime.datetime]:
+        
+        track_states = extract_states(tracks_set)
+        truth_states = extract_states(truth_set)
+        timestamps = sorted({
+            state.timestamp
+            for state in chain(track_states, truth_states)})
+        return timestamps
+
+
+def extract_states(object_with_states, return_ids=False):
```

Comment:
We can combine the 3 "copies" of this function in the codebase into one at some point...